core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/TupleCodec.java

```diff
@@ -161,70 +161,90 @@ public TupleValue parse(@Nullable String value) {
       throw new IllegalArgumentException(
           String.format(
-              "Cannot parse tuple value from \"%s\" at character %d: expecting '(' but got '%c'",
+              "Cannot parse tuple value from \"%s\", at character %d expecting '(' but got '%c'",
               value, position, value.charAt(position)));
     }
 
     position++;
     position = ParseUtils.skipSpaces(value, position);
 
-    if (value.charAt(position) == ')') {
-      return tuple;
-    }
-
     CodecRegistry registry = cqlType.getAttachmentPoint().getCodecRegistry();
 
-    int i = 0;
-    while (position < value.length()) {
+    int field = 0;
+    while (position < length) {
+      if (value.charAt(position) == ')') {
+        position++;
+        if (position == length) {
+          return tuple;
+        } else {
+          position = ParseUtils.skipSpaces(value, position);
+          if (position == length) {
+            return tuple;
+          }
+          throw new IllegalArgumentException(
+              String.format(
+                  "Cannot parse tuple value from \"%s\", at character %d expecting EOF or blank, but got \"%s\"",
+                  value, position, value.substring(position)));
+        }
+      }
       int n;
       try {
         n = ParseUtils.skipCQLValue(value, position);
       } catch (IllegalArgumentException e) {
         throw new IllegalArgumentException(
             String.format(
-                "Cannot parse tuple value from \"%s\", invalid CQL value at character %d: %s",
-                value, position, e.getMessage()),
+                "Cannot parse tuple value from \"%s\", invalid CQL value at field %d (character %d)",
+                value, field, position),
             e);
       }
 
       String fieldValue = value.substring(position, n);
-      DataType elementType = cqlType.getComponentTypes().get(i);
+      DataType elementType = cqlType.getComponentTypes().get(field);
       TypeCodec<Object> codec = registry.codecFor(elementType);
       Object parsed;
       try {
         parsed = codec.parse(fieldValue);
       } catch (Exception e) {
         throw new IllegalArgumentException(
             String.format(
-                "Cannot parse tuple value from \"%s\", invalid CQL value at character %d: %s",
-                value, position, e.getMessage()),
+                "Cannot parse tuple value from \"%s\", invalid CQL value at field %d (character %d): %s",
+                value, field, position, e.getMessage()),
             e);
       }
-      tuple = tuple.set(i, parsed, codec);
+      tuple = tuple.set(field, parsed, codec);
 
       position = n;
-      i += 1;
 
       position = ParseUtils.skipSpaces(value, position);
+      if (position == length) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Cannot parse tuple value from \"%s\", at field %d (character %d) expecting ',' or ')', but got EOF",
+                value, field, position));
```

Comment:
Maybe you could just break here, in order to hit the `throw` statement after the loop. It makes sense since `position == length` is the exit condition of the loop.

Reply:
But the error message would have to be adapted since here we are expecting a comma or the closing `)` and at the other location we are expecting a CQL literal or `)`. I think it would add complexity to factorize the two error messages into one `throw` statement.

Reply:
I hadn't noticed the difference. Makes sense then 👍